### ISSUE TITLE
[API Proposal]: Consider adding JsonDocument.TryParse(), Take 2

### ISSUE THREAD

**Issue by chenxinyanc** [dotnet/runtime#82605]
### Background and motivation

Today I want to implement a function that tries to parse a certain HTTP header for a certain JSON property.

```c#
public string? TryGetCertainProperty()
{
    string? headerValue = httpRequest.Headers["certain-header"].FirstOrDefault();
    if (headerValue == null) return null;
    JsonDocument doc;

    try
    {
        // ???
        // https://github.com/dotnet/runtime/issues/387
        doc = JsonDocument.Parse(headerValue);
    }
    catch (Exception)
    {
        return null;
    }

    if (doc.RootElement.ValueKind != JsonValueKind.Object) return null;
    if (!doc.RootElement.TryGetProperty("CertainProperty", out var el)) return null;
    if (el.ValueKind != JsonValueKind.String) return null;
    return el.GetString();
}
```

I think something might be missing.

I'm opening this issue because the previous one, #387, was closed possibly due to unclear requirements.

### API Proposal

```csharp
namespace System.Text.Json;

public class JsonDocument
{
    /// <summary>
    /// Tries to parse a JsonDocument instance from the given string containing the JSON expression.
    /// </summary>
    /// <param name="json">a string that possibly contain the JSON expression.</param>
    /// <param name="options">parsing options.</param>
    /// <returns>parsed JSON document, or <c>null</c> if the provided string cannot be parsed as JSON document.</returns>
    /// <exception cref="ArgumentNullException">json is <c>null</c>.</exception>
    public static JsonDocument? TryParse(string json, JsonDocumentOptions options = default);
    public static JsonDocument? TryParse(ReadOnlyMemory<char> json, JsonDocumentOptions options = default);
    public static JsonDocument? TryParse(ReadOnlySpan<char> json, JsonDocumentOptions options = default);
}
```

### API Usage

```csharp
var doc = JsonDocument.Parse("{\"foo\": 123}");     // JsonDocument
var doc = JsonDocument.Parse("foo");        // null
var doc = JsonDocument.Parse("{\"foo");     // null
var doc = JsonDocument.Parse("");           // null
var doc = JsonDocument.Parse(null);         // ArgumentNullException
```

### Alternative Designs

```csharp
namespace System.Text.Json;

public class JsonDocument
{
    /// <summary>
    /// Tries to parse a JsonDocument instance from the given string containing the JSON expression.
    /// </summary>
    /// <param name="json">a string that possibly contain the JSON expression.</param>
    /// <param name="doc">a variable to receive parsed JsonDocument.</param>
    /// <param name="options">parsing options.</param>
    /// <returns><c>true</c> if the parsing is successful; <c>false</c> otherwise.</returns>
    /// <exception cref="ArgumentNullException">json is <c>null</c>.</exception>
    public static bool TryParse(string json, [NotNullWhen(true)] out JsonDocument? doc, JsonDocumentOptions options = default);
}
```

### Risks

No breaking changes or performance regression as I'm aware of so far,

**Comment by chenxinyanc** [comment]
Per [Framework design guidelines >> Exceptions and Performance # Try-Parse Pattern](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exceptions-and-performance#try-parse-pattern), the only goal to introduce such `Try...` methods is, to avoid throwing `Exception` when the caller already knows the parsing can fail, so as to save the performance penalty of throwing an `Exception`.

> For extremely performance-sensitive APIs, an even faster pattern than the Tester-Doer Pattern described in the previous section should be used. 

Personally, I would doubt whether "an API could fail due to different failure conditions" could be the reason why we don't have such a `TryParse` method that *only throws for `null` inputs* at all.

I've seen from previous discussion that "building a mode in the reader that doesn't throw on invalid JSON will slow it down a bit". This is totally understandable, but I'd argue whether it can be the main reason why we don't have such public API
* In most of the cases, I just don't care the 1% (or sometimes even 10%) performance penalty caused by `TryParse`. (10% of 1ms is 0.1ms per HTTP request). Instead it's frustrating not having a `TryParse` function and I'm being forced to write a `catch` clause that just returns null.
* I don't suppose parsing JSON in this mode would be slower than parsing an JSON and throw an Exception.
* It is always possible to make some optimization inside this `TryParse` implementation without breaking this public API surface, on condition that such performance penalty is prominent enough.
* Even if I'm concerned about such performance penalty, after carefully profiling, I might revert back to the `try... Parse .. catch ... return null` approach, but at that point of time, I will have a very strong justification to write code that looks unintuitive, and will attach my justification to the comments.

**Comment by ChrML** [comment]
I think a valid question to ask is whether an invalid JSON is a common enough scenario to be worth optimizing for. By example, all controllers that use [FromBody] with JSON formatted data can throw an exception if the format is not as expected, but most of the time they are called correctly. Apparently the hit of raising that exception is not large enough for that to be considered a DDoS vector.

I would guess that the header in question here is a custom header, or that some simple checks such as String.Contains could be run first to make it more probable it's json. And that would account for 99.99% of requests.

**Comment by eiriktsarpalis** [comment]
> * In most of the cases, I just don't care the 1% (or sometimes even 10%) performance penalty caused by `TryParse`. (10% of 1ms is 0.1ms per HTTP request). Instead it's frustrating not having a `TryParse` function and I'm being forced to write a `catch` clause that just returns null.

If performance isn't an issue for you, wouldn't a helper method that just catches exceptions like you describe address your use case?

> I don't suppose parsing JSON in this mode would be slower than parsing an JSON and throw an Exception.

The primary concern isn't comparing performance of a `TryParse` method that returns false and throwing exceptions. It's mostly an issue of not regressing the happy path of parsing valid JSON. The `JsonDocument.Parse` implementation relies heavily on the `Utf8JsonReader` type which is generally speaking designed to throw when invalid data is encountered, mostly because it is optimizing for successful reads. Changing that would require some redesign, and I'm not sure that it wouldn't result in regressions for the 99% use case.

Duplicate of #387. We won't be adding such a method since it violates guidelines for `Try-` methods, namely a `false` result should represent a single failure mode. We might however want to consider an `IsValidJson` method that simply checks if a piece of data is valid JSON, so that the numerous `Parse` methods in our DOM implementations are guaranteed to succeed if that check returns true.

**Comment by TheGuy920** [comment]
I am personally experiencing performance issues with large amounts of JSON parsing. My use case involves traversing a large tree with hundreds of files, where JSON content can be present in files with any naming scheme. To handle this, my current solution involves parsing all files with some basic pre-parse filtering within a try/catch block. 

Performance is a top priority as I need to parse large amounts of data into a `JsonDocument`. The existing approach of using try/catch for invalid JSON is not optimal due to the overhead of exceptions.

While a `IsValidJson` method is *a* solution to the problem, we then run into the issue of needing to reread the data which will create another performance issue. The ideal solution would be a single pass with a non-exception exit upon failure to avoid all performance issues and ignore invalid JSON data. 

Now when parsing into a class, I understand the problem of a `TryParse` however any/all valid JSON is supposed to be able to parse into a `JsonDocument`, no? Meaning the boolean return case would represent a `Valid` or `Invalid` json data state? When parsing into a "dynamic" object of sorts, there are no constraints which would present additional errors. This would mean the only presentable error when parsing into a non-structured object like `JsonDocument` would be like a syntax error, right? 

So thats what I am looking for. The ability to parse large amount of unknown data into a `JsonDocument` if applicable while skipping invalid data (syntax error) without any performance issues that come from processing exceptions. I hope I am understanding this all correctly? I've been using C# for many years but I am all self taught so I dont usually get everything right first try

**Comment by ChrML** [comment]
@TheGuy920 Does the exceptions being thrown really show up as impactful? I'd imagine the extra time for that would be a lot less impactful compared to the IO time of actually reading the files. Exceptions are slower than results, but exceptions shouldn't be that slow.

**Comment by TheGuy920** [comment]
Yea it is that impactful, at least when I am debugging. It so impactful I have yet to benchmark it. There is a noticeable time difference of 10+ seconds but just to be sure, I will go ahead and use the `StopWatch` for some preliminary benchmarking so we can get a rough idea.
 ```csharp
 Stopwatch entireBatch = Stopwatch.StartNew();
 ConcurrentBag<(bool Status, string Name, long FileSize, TimeSpan ParseTime)> values = [];
 Parallel.ForEach(this.ProjectDirectory.EnumerateFiles("*.*", SearchOption.AllDirectories), file =>
{
    /* pre-parse filtering */

    using FileStream fstream = file.OpenRead();
    Stopwatch sw = Stopwatch.StartNew();

    try
    {
        var jdoc = JsonSerializer.Deserialize<JsonDocument>(fstream, JsonParseOptions);
        values.Add((false, file.Name, fstream.Length, sw.Elapsed));

        // processing
        var root = jdoc.RootElement;
    }
    catch { values.Add((true, file.Name, fstream.Length, sw.Elapsed)); }
});

entireBatch.Stop();
```
Keep in mind the failed parsing attempts is due to the input/file data not being JSON related at all

* Release Mode + Debugger Attached
```
Warning: 0 : Entire Batch Time: 00:00:07.4868945
Warning: 0 : Failed Items: 192 | Average Parse Time: 444.792... μs | Average File Size: 466059 bytes
Warning: 0 : Success Items: 58 | Average Parse Time: 441.379... μs | Average File Size: 8749 bytes
```
* Release Mode + No Debugger Attached (saved to log file)
```
Entire Batch Time: 00:00:00.0426526
Failed Items: 192 | Average Parse Time: 454.688... μs | Average File Size: 466059 bytes
Success Items: 58 | Average Parse Time: 444.828... μs | Average File Size: 8749 bytes
```
Having the debugger attached was not an issue at all until it came to JSON deserialization. The JSON deserialization should fail within the first few bytes as its attempting to parse non-json yet it somehow takes the same amount of time or longer for parsing valid json into JsonDocument objects? I'm no expert when it comes to stuff like this, and maybe I am doing something wrong, but its very annoying to run this in debugging mode with a debugger attached with a 7 second difference

**Comment by ChrML** [comment]
@TheGuy920 Note that your stopwatch also includes the time it takes to read the bytes from the disk. You start it after OpenRead, but it's a stream, so the bytes are actually read from the disk by Deserialize. Your code as it is now is better for performance, but if you want a bench on the deserialization only, you need to read the Stream to a string first before starting stopwatch and pass the string to Deserialize.

Having the debugger attached makes a huge difference when it comes to exceptions. Probably because VS is given an opportunity to intercept it. In release mode, exceptions are relatively fast. Benchmarks I've seen suggest you can throw/catch around 30 000 of them per second +- depending on callstack depth and rethrows on the way. That suggests exceptions on release mode without debugger should not contribute to more than ~0.1 sek per 1000 invalid file you read (probably less than that too).

As it's written above by others, having a Try .. path is likely to make JSON serialization of valid data a bit slower due to various errorcode checks and extra results to pass around internally. It's good that it's optimized for valid data I think.

I don't really think the exceptions are making a huge impact here tho.

In the upcoming NET9 release, exceptions are optimized quite a bit. Reason for that is that people have realized exceptions are not avoidable even in high-load applications. Of course you'd avoid throwing them everywhere as results to non-exceptional stuff. But as soon as a connection is aborted/lost or similar, they happen and there are not many good alternatives.

**Comment by TheGuy920** [comment]
So I guess the real killer is the VS debugger then? When I am running I can see in the output window that VS is bombarded with JsonExceptions regardless of the try { } catch { } so my idea to fix this (debugger performance issues) was for an exceptionless exit for the parse/deserialize. Idk. Thats my problem and I don't know of a solution for it.

**Comment by ChrML** [comment]
@TheGuy920 Not sure either if VS can be configured to handle this faster. Performance during debugging is usually not an issue, as a debugger is rarely attached in a production scenario where the performance actually matters. Usually, only unit-tests and specific issues during dev are debugged.

**Comment by TheGuy920** [comment]
yea well... a few milliseconds versus 7 seconds is not small, its quite annoying and impedes how quickly I can progress and debug

**Comment by ChrML** [comment]
@TheGuy920 Use the debugger on specific problems and a subset/examples/unit-tests of your data, not on the full application scope. E.g. if there is a problem with a file, run it just with that file, or better, make a test with that sample of data. Saves you much time debugging out-of-scope stuff.

Personally it's been years since I've used the debugger to run everything. I only use it on specific unit-tests after I've written one that uncovers the problem I wanted to fix, if it's not obvious at once what causes the problem. Occasionally I use it for profiling too, but with a dataset limited to what I want to look closer at.

**Comment by eiriktsarpalis** [comment]
Per feedback in #387 it is unlikely we would ever add a `JsonDocument.TryParse` because it would require nontrivial changes to the underlying `Utf8JsonReader` type in ways that would most likely compromise performance for everyone. Another reason is that `JsonDocument` is `IDisposable`, so returning one using an `out` parameter makes disposal impractical or easy to forget.

It's more likely that we could consider an `IsValidJson` helper, although that approach comes with its own trade-offs (duplicate parsing, not usable when streaming).
